openstackmap.py

```diff
@@ -24,245 +24,480 @@
-    global vms
-    global vms_by_id
-    vms_by_id = dict()
-
-    if not os.path.exists(pers_region + '/vms.pickle'):
-        refresh_nova()
-    else:
-        vms = load('vms') 
-
-    vms_by_id = dict( (vm['id'], vm) for vm in vms)
-
-def refresh_cinder():
-    global volumes, backup_volumes, snapshot_cinder    
-
-    cinder = osclients.get_cinderclient()
-
-    with open(pers_region + '/volumes.pickle', 'wb') as f:
-        volumes = list(volume.__dict__ for volume in cinder.volumes.list(
-            search_opts={'all_tenants': 1}))
-        pickle.dump(volumes, f, protocol=-1)
-
-def load_cinder():
-    global volumes
-    global volumes_by_id
-
-    if not os.path.exists(pers_region + '/volumes.pickle'):
-        refresh_cinder()
-    else:
-        volumes = load('volumes')
-
-    volumes_by_id = dict( (volume['id'], volume) for volume in volumes)
-
-def refresh_glance():
-    global images
-    
-    glance = osclients.get_glanceclient()
-    with open(pers_region + '/images.pickle', 'wb') as f:
-        images = list(image.to_dict() for image in  glance.images.findall())
-        pickle.dump(images, f, protocol=-1)
-
-def load_glance():
-    global images, images_by_id
-    if not os.path.exists(pers_region + '/images.pickle'):
-        refresh_glance()
-    else:
-        images = load('images')
-    images_by_id = dict( (image['id'], image) for image in images)
-
-def load_all():
-    load_keystone()
-    load_neutron()
-    load_nova()
-    load_glance()
-    load_cinder()
-
-def refresh_all(also_keystone=True):
-    if also_keystone:
-        refresh_keystone()
-    refresh_neutron()
-    refresh_nova()
-    refresh_glance()
-    refresh_cinder()
-
-def change_region(new_region):
-    global pers_region
-    pers_region = persistence_dir + '/' + new_region
-    if not os.path.exists(pers_region):
-        os.mkdir(pers_region)
-    os.environ['OS_REGION_NAME'] = new_region
-    load_all()
-
-if not os.path.exists(persistence_dir):
-    os.mkdir(persistence_dir)
-
-if not os.path.exists(pers_keystone):
-    os.mkdir(pers_keystone)
-
-if not os.path.exists(pers_region):
-    os.mkdir(pers_region)
 
+        nets = dict((network['id'], network) for network in networks)
+        snets = dict((subnet['id'], subnet) for subnet in subnets)
+        routers = dict((router['id'], router) for router in routers)
+        floatingips = dict((floatingip['id'], floatingip) for floatingip in
+                               floatingips)
+        sec_grps = dict((sg['id'], sg) for sg in sec_grps)
+        ports = dict((port['id'], port) for port in ports)
+
+        if dict_object:
+            # neutron objects are already dicts, so they do not need conversion
+            pass
+        else:
+            self.networks = nets
+            self.subnets = snets
+            self.routers = routers
+            self.floatingips = floatingips
+            self.security_groups = sec_grps
+            self.ports = ports
+            return
```

Comment:
Yes, it is correct. If dict_object is false is because strategy_objects is DIRECT_OBJECTS. Persistence is not available with direct objects, only when using the other strategy types that implies converting the object to a dictionary. This is because the direct objects as returned by the API have a lot of internal objects that cannot be persisted.